pkg/extensions/sync/destination.go

```diff
@@ -80,116 +81,132 @@ func (registry *DestinationRegistry) CanSkipImage(repo, tag string, imageDigest
+	tempImageStore := getImageStoreFromImageReference(repo, imageReference, registry.log)
 
 	defer os.RemoveAll(tempImageStore.RootDir())
 
-	registry.log.Info().Str("syncTempDir", path.Join(tempImageStore.RootDir(), repo)).Str("reference", reference).
+	registry.log.Info().Str("syncTempDir", path.Join(tempImageStore.RootDir(), repo)).Str("repository", repo).
 		Msg("pushing synced local image to local registry")
 
 	var lockLatency time.Time
 
-	manifestBlob, manifestDigest, mediaType, err := tempImageStore.GetImageManifest(repo, reference)
+	index, err := storageCommon.GetIndex(tempImageStore, repo, registry.log)
 	if err != nil {
 		registry.log.Error().Str("errorType", common.TypeOf(err)).
-			Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).Str("reference", reference).
-			Msg("couldn't find synced manifest in temporary sync dir")
+			Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).
+			Msg("failed to get repo index from temp sync dir")
 
 		return err
 	}
 
-	// is image manifest
-	switch mediaType {
-	case ispec.MediaTypeImageManifest:
-		if err := registry.copyManifest(repo, manifestBlob, reference, tempImageStore); err != nil {
-			if errors.Is(err, zerr.ErrImageLintAnnotations) {
-				registry.log.Error().Str("errorType", common.TypeOf(err)).
-					Err(err).Msg("couldn't upload manifest because of missing annotations")
+	for _, desc := range index.Manifests {
+		reference := GetDescriptorReference(desc)
 
-				return nil
-			}
+		manifestBlob, manifestDigest, mediaType, err := tempImageStore.GetImageManifest(repo, reference)
+		if err != nil {
+			registry.log.Error().Str("errorType", common.TypeOf(err)).
+				Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).Str("repo", repo).Str("reference", reference).
+				Msg("failed to get manifest from temporary sync dir")
 
 			return err
 		}
-	case ispec.MediaTypeImageIndex:
-		// is image index
-		var indexManifest ispec.Index
 
-		if err := json.Unmarshal(manifestBlob, &indexManifest); err != nil {
-			registry.log.Error().Str("errorType", common.TypeOf(err)).
-				Err(err).Str("dir", path.Join(tempImageStore.RootDir(), repo)).
-				Msg("invalid JSON")
+		// is image manifest
+		switch mediaType {
+		case ispec.MediaTypeImageManifest:
+			if err := registry.copyManifest(repo, manifestBlob, reference, tempImageStore); err != nil {
+				if errors.Is(err, zerr.ErrImageLintAnnotations) {
+					registry.log.Error().Str("errorType", common.TypeOf(err)).
+						Err(err).Msg("failed to upload manifest because of missing annotations")
 
-			return err
-		}
+					return nil
+				}
 
-		for _, manifest := range indexManifest.Manifests {
-			tempImageStore.RLock(&lockLatency)
-			manifestBuf, err := tempImageStore.GetBlobContent(repo, manifest.Digest)
-			tempImageStore.RUnlock(&lockLatency)
+				return err
+			}
+		case ispec.MediaTypeImageIndex:
```

Comment:
Does this over indexes referencing indexes?

Reply:
wouldn't all be referenced in index.json though?

Reply:
updated logic and added test.